src/sink/nats.rs

```diff
@@ -0,0 +1,324 @@
+    reply_channel: Sender<sink::Reply>,
+    connection: Option<NatsConnection>,
+    error_rx: Receiver<()>,
+    error_tx: Sender<()>,
+    merged_meta: OpMeta,
+}
+
+impl offramp::Impl for Nats {
+    fn from_config(config: &Option<OpConfig>) -> Result<Box<dyn Offramp>> {
+        if let Some(config) = config {
+            let config: Config = Config::new(config)?;
+            let (dummy_tx, _) = bounded(1);
+            let (error_tx, error_rx) = bounded(crate::QSIZE);
+            Ok(SinkManager::new_box(Self {
+                sink_url: TremorUrl::from_offramp_id("nats")?,
+                config,
+                postprocessors: vec![],
+                reply_channel: dummy_tx,
+                connection: None,
+                error_rx,
+                error_tx,
+                merged_meta: OpMeta::default(),
+            }))
+        } else {
+            Err("Nats offramp requires a configuration.".into())
+        }
+    }
+}
+
+impl Nats {
+    fn handle_connection(&mut self) -> Result<Option<&NatsConnection>> {
+        while let Ok(()) = self.error_rx.try_recv() {
+            self.connection = None;
+        }
+        if self.connection.is_none() {
+            self.connection = Some(self.config.connection()?);
+            return Ok(self.connection.as_ref());
+        }
+        Ok(None)
+    }
+}
+
+#[async_trait::async_trait]
+impl Sink for Nats {
+    #[allow(clippy::cast_possible_truncation)]
+    async fn on_event(
+        &mut self,
+        _input: &str,
+        codec: &mut dyn Codec,
+        _codec_map: &HashMap<String, Box<dyn Codec>>,
+        mut event: Event,
+    ) -> ResultVec {
+        self.handle_connection()?;
+        let ingest_ns = event.ingest_ns;
+        let processing_start = Instant::now();
+        let op_meta = &event.op_meta;
+        self.merged_meta.merge(op_meta.clone());
+        let insight_event = event.insight_ack();
+        if let Some(connection) = &mut self.connection {
+            for (value, meta) in event.value_meta_iter() {
+                let encoded = codec.encode(value)?;
+                let processed =
+                    postprocess(self.postprocessors.as_mut_slice(), ingest_ns, encoded)?;
+                let headers = meta.get("nats").and_then(|v| v.get_object("headers"));
+                let reply = meta.get("nats").and_then(|v| v.get_str("reply"));
+                for payload in processed {
+                    // prepare message reply
+                    let mut message_reply = None;
+                    if let Some(reply) = reply {
+                        message_reply = Some(reply);
+                    } else if let Some(config_reply) = &self.config.reply {
+                        message_reply = Some(config_reply.as_str());
+                    }
+                    // prepare message headers
+                    let mut key_val: Vec<(&str, &str)> =
+                        Vec::with_capacity(self.config.headers.len());
```

Comment:
It would be possible including the length of custom headers (+/- the map function) :
```suggestion
                    let mut key_val: Vec<(&str, &str)> =
                        Vec::with_capacity(self.config.headers.len() + headers.map(HashMap::len).unwrap_or_default());
```